glue-codes/openfast-cpp/src/OpenFAST.cpp

```diff
@@ -338,584 +757,622 @@ void fast::OpenFAST::calc_nacelle_force(
-  forceNodeVel.resize(nTurbinesProc);
-  
-  for (int iTurb=0; iTurb < nTurbinesProc; iTurb++) {
-    
-    TurbineBasePos[iTurb].resize(3);
-
-    int globProc = turbineMapProcToGlob[iTurb];
-    TurbID[iTurb] = globTurbineData[globProc].TurbID;
-    FASTInputFileName[iTurb] = globTurbineData[globProc].FASTInputFileName ;
-    CheckpointFileRoot[iTurb] = globTurbineData[globProc].FASTRestartFileName ;
-    for(int i=0;i<3;i++) {
-      TurbineBasePos[iTurb][i] = globTurbineData[globProc].TurbineBasePos[i];
+    
+    
+    // Construct a group containing all procs running atleast 1 turbine in FAST
+    MPI_Group_incl(worldMPIGroup, nProcsWithTurbines, &turbineProcs[0], &fastMPIGroup) ;
+    int fastMPIcommTag = MPI_Comm_create(mpiComm, fastMPIGroup, &fastMPIComm);
+    if (MPI_COMM_NULL != fastMPIComm) {
+        MPI_Comm_rank(fastMPIComm, &fastMPIRank);
     }
-    numForcePtsBlade[iTurb] = globTurbineData[globProc].numForcePtsBlade;
-    numForcePtsTwr[iTurb] = globTurbineData[globProc].numForcePtsTwr;
-    nacelle_cd[iTurb] = globTurbineData[globProc].nacelle_cd;
-    nacelle_area[iTurb] = globTurbineData[globProc].nacelle_area;
-    air_density[iTurb] = globTurbineData[globProc].air_density;
-
-  }
-
-  // Allocate memory for Turbine datastructure for all turbines
-  FAST_AllocateTurbines(&nTurbinesProc, &ErrStat, ErrMsg);
-  
-  // Allocate memory for OpFM Input types in FAST
-  cDriver_Input_from_FAST.resize(nTurbinesProc) ;
-  cDriver_Output_to_FAST.resize(nTurbinesProc) ;
-  
-  cDriverSC_Input_from_FAST.resize(nTurbinesProc) ;
-  cDriverSC_Output_to_FAST.resize(nTurbinesProc) ;
 
+    turbineData.resize(nTurbinesProc);
+    velForceNodeData.resize(nTurbinesProc);
+    
+    for (int iTurb=0; iTurb < nTurbinesProc; iTurb++) {
+        
+        turbineData[iTurb].TurbineBasePos.resize(3);
+        
+        int iTurbGlob = turbineMapProcToGlob[iTurb];
+        turbineData[iTurb].TurbID = globTurbineData[iTurbGlob].TurbID;
+        turbineData[iTurb].FASTInputFileName = globTurbineData[iTurbGlob].FASTInputFileName ;
+        turbineData[iTurb].FASTRestartFileName = globTurbineData[iTurbGlob].FASTRestartFileName ;
+        for(int i=0;i<3;i++)
+            turbineData[iTurb].TurbineBasePos[i] = globTurbineData[iTurbGlob].TurbineBasePos[i];
+        turbineData[iTurb].numForcePtsBlade = globTurbineData[iTurbGlob].numForcePtsBlade;
+        turbineData[iTurb].numForcePtsTwr = globTurbineData[iTurbGlob].numForcePtsTwr;
+        
+        velForceNodeData[iTurb].resize(4); // To hold data for 4 time steps
+        
+    }
+
+    // Allocate memory for Turbine datastructure for all turbines
+    FAST_AllocateTurbines(&nTurbinesProc, &ErrStat, ErrMsg);
+
+    // Allocate memory for ExtInfw Input types in FAST
+    i_f_FAST.resize(nTurbinesProc) ;
+    o_t_FAST.resize(nTurbinesProc) ;
+    
+    sc_i_f_FAST.resize(nTurbinesProc) ;
+    sc_o_t_FAST.resize(nTurbinesProc) ;
+    
+}
+
+void fast::OpenFAST::allocateMemory2(int iTurbLoc) {
```

Comment:
What's the difference between `allocateMemory` and `allocateMemory2`, give it a descriptive name may be?

Reply:
Thanks. Changed names to `allocateMemory_preInit()` and `alllocateMemory_postInit()`. Implemented in 9f34f9eb119d38dc0c78bfe58fdd23d084d02430.